### ISSUE TITLE
Optimiser interface is not documented

### ISSUE THREAD

**Issue by purzelrakete** [FluxML/Flux.jl#836]
The optimiser api is not documented anywhere. Here are some questions that came up for me:

- the relationship between `apply!` and `update!` is not clear just based on the names of these methods. It appears that `apply!` updates gradients, while `update!` updates parameters.
- it is not immediately obvious what `apply!` should mutate. Maybe just the optimiser? This is exacerbated by the fact that it returns `Δ`.  Is there a reason to return `Δ` when it is already being mutated? Some light documentation would quickly dispel any confusion.
- There is no documentation. I looked at the pull request for the optimiser changes as well as the commit message and the code itself. 
- Parameters are named `x`. This is not what I would expect, since x usually refers to the features. This could quickly be cleared up with some docstrings. 
- It is not immediately clear why parameters are being passed into `apply!`. It seems that we are not allowed to mutate `x`, and it also isn't used by any of the optimisers to compute gradient updates. It would help to document the the lifecycle of the `apply!` method to understand that we need to attach optimiser state to each individual parameter. The reason is that `apply!` can get called many times in the same update step, once for each parameter. Some documentation would be helpful. 
- The docstrings are not in sync with the coder, or not appropriate for the methods being described. Example: "Classic gradient descent optimiser with learning rate `η`. For each parameter `p` and its gradient `δp`, this runs `p -= η*δp`.". Names of the parameters don't match the code here, which uses `x` and `Δ` instead of `p` and `δp`. 

I think it might be helpful to explicity write out an interface along with some documentation. Then each Optimiser will need less documentation. Ignoring the fact that `Optimiser` already exists, something like this maybe:

```julia:
"""
Apply changes to gradients. 

Optimisers are called once for each parameter in each update step. Because of this, the parameter `x` is also passed to each `apply!` call. Optimisers can then do bookkeeping by namespacing changes to that parameter only, e.g. by keeping an internal dictionary of parameters to internal counters.
"""
abstract type Optimiser; end

"""
Applies changes to the gradient.

# Arguments
- `o`: the optimiser. Keep optimiser state in here and update on apply! as appropriate.
- `x`: the parameter being optimised. It may not be changed by apply.
- `Δ`: the gradients w.r.t the parameter. Gradients should be mutated by apply using elementwise operations.

# Returns
- `Δ`: the updated gradients w.r.t the parameter.
"""
apply!(o::Optimiser, x, Δ) = error("not implemented")

"""
    Descent(η)

Classic gradient descent optimiser with learning rate `η`.  When applied, the given gradient `Δ` is mutated to be equal to `ηΔ`.
"""
struct Descent <: Optimiser
  eta::Float64
end

Descent() = Descent(0.1)
apply!(o::Descent, x, Δ) = Δ .*= o.eta
```
